packages/braid-design-system/lib/components/private/buildDataAttributes.ts

```diff
@@ -1,6 +1,68 @@
+import dedent from 'dedent';
+
 export type DataAttributeMap = Record<string, string | number>;
 
-export default (data: DataAttributeMap = {}): DataAttributeMap => {
+interface DataAttributeParams {
+  data?: DataAttributeMap;
+  validateRestProps: Record<string, any> | false;
+}
+
+export default ({ data, validateRestProps }: DataAttributeParams) => {
+  if (process.env.NODE_ENV !== 'production') {
+    if (validateRestProps !== false) {
+      const dataAttrs = Object.keys(validateRestProps).filter((prop) =>
+        prop.startsWith('data-'),
+      );
+
+      const formatter = (v: boolean | string | number) => {
+        if (typeof v === 'boolean') {
+          return v ? '' : `={false}`;
+        }
+
+        if (typeof v === 'number') {
+          return `={${v}}`;
+        }
+
+        return `="${v}"`;
+      };
+
+      if (dataAttrs.length > 0) {
+        // eslint-disable-next-line no-console
+        console.warn(
+          dedent`
+          Braid components do not support the native data attribute format. Use the “data” prop instead.
```

Comment:
Would it be worth linking to some docs here too? https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes